### PR TITLE
Add snapml solver

### LINKED ISSUE #16: add snapml solver
https://snapml.readthedocs.io/en/latest/logistic_regression.html

### PR DESCRIPTION
**Pull request by tanglef**
Fixes #16
Also add SUSY dataset in `libsvm` (very high number of samples, very few features) as it is one of the examples in snapml doc to motivate the use of the package.
I can run a `publish` command with news20, SUSY and simulated datasets if you want

### CODE REVIEW

**Comment by agramfort** [comment]
looks good ! do you get meaningful results?

**Comment by mathurinm** [comment]
> very high number of samples, very few features

with L1 penalty ? :thinking: 

EDIT: it's (5000000, 18). IMO it may be counterproductive to include it in the benchmark. In this setting one should go towards very specific methods (greedy CD?) no ?
Do you know why they advertise it @tanglef ?

**Comment by tanglef** [comment]
It's used with L2 penalty (PR coming very soon), but the solver is also faster in this setting (but I agree)
In fact, all of their examples are with L2 penalty

**Comment by tanglef** [comment]
Converted to draft as MacOs seem to be problematic in tests

**Comment by mathurinm** [comment]
@tanglef I did not see the tests running for this PR, where do you see them ?

**Comment by tanglef** [comment]
It turned out to be red on my fork.
And it broke because we need a `brew install libomp`-like, but there is no conda package for that for MacOS (as far as I know and found out).
I'm still not sure how to solve this for now. But other tests are green btw

**Comment by agramfort** [comment]
do we know why CIs don't run here?

otherwise can you try installing openmp with:

https://anaconda.org/conda-forge/openmp

?

**Comment by tanglef** [comment]
There was an issue with github CI, and then I turned it into a draft to try a few things on the side.

Also, about the results, nothing exceptional with the L1 penalty, seems to be outperformed by another solver everytime, but a generally not bad.
With the L2 penalty, there seems to be a good speedup sometimes against sklearn/cd.

When this CI will go through with OSX I'll do the other one for L2 logreg.

**Comment by agramfort** [comment]
can you share a figure?

**Comment by tanglef** [comment]
On L1 penalty:

| news20 | SUZY | simulated |
|---|---|---|
| Capture d’écran de 2022-03-16 23-08-49|Capture d’écran de 2022-03-16 23-09-03 |Capture d’écran de 2022-03-16 23-11-23 |

**Comment by tanglef** [comment]
With L2 penalty:
|news20 | news20 | SUZY |
| ---| --- | ---|
|Capture d’écran de 2022-03-16 23-16-46 |Capture d’écran de 2022-03-16 23-16-52 |Capture d’écran de 2022-03-16 23-17-46|

**Comment by agramfort** [comment]
now it does not crash but it does not complete ... 🤦 

I would see next if openmp is the pb even without snapml or not.

@tanglef can you investigate?

**Comment by agramfort** [comment]
maybe try setting

export OMP_NUM_THREADS=1

in the CI?

**Comment by tanglef** [comment]
Will try (I run some of the tests on my fork too :) but by default the number of threads used by snapml is 1)

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -51,6 +51,8 @@ jobs:
       run: |
         brew install libomp
         ls /usr/local/opt/libomp/lib/
+        export OMP_NUM_THREADS=1
```
Comment (agramfort) [inline .github/workflows/main.yml]: not sure this survives between the different run blocks. I would put this just above the pytest

**Comment by tanglef** [comment]
Update: The issue happens in the 2nd part of the test when there is the `--skip-install` flag.

**Comment by agramfort** [comment]
can you replicate locally?

**Comment by tanglef** [comment]
No, I'm under linux (it all works) and I don't have a mac. I only run multiple CIs on github

**Comment by tanglef** [comment]
So, the CI is green now, but I'm not sure if it's actually good or just because github actions were down for a few minutes. I'll run a trigger in a few minutes to be sure

**Comment by agramfort** [comment]
can you cleanup and see if CIs are green? If so maybe we're good...

**Comment by tanglef** [comment]
CI is green and files are cleaned. Thanks a lot !
Btw, if you have an explanation about why the need to specify the number of threads... I'd be glad to know more, right now it's a little "a black magic trick"-like to me

**Review by agramfort — APPROVED** [review]
@tomMoral @mathurinm the floor is yours !

thx @tanglef

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -58,6 +63,8 @@ jobs:
 
     - name: Test
       run: |
+        export OMP_NUM_THREADS=1
```
Comment (tomMoral) [inline .github/workflows/main.yml:66]: Can you create an issue that reports the failure so we can try to fix this later?

For a bit of explainations: openMP is a library that is used to make some parallel computations, typically for numpy operations but also probably for `snapml`. Default behavior is to use all cores to run the operations in parallel.

By setting `OMP_NUM_THREADS=1`, we bypass the default behavior and requires sequential computations. So if setting this solves the problem, this is most probably an issue of concurrency in `snapml` code.

**Review by tomMoral — APPROVED** [review]
LGTM!

@tanglef can you report the issue so we keep tack of this. The `OMP_NUM_THREADS=1` trick is not really the normal behavior, so it is  good enough for this PR but it probably means that there is some underlying issues for OSX users.

File: .github/workflows/main.yml (hunk 2 above)
Comment (tanglef) [reply .github/workflows/main.yml:66]: Yes, that's done now (#20)
Ok, thanks for the explanation

**Pull request merged**